File: rust/src/context.rs
use std::sync::Arc;
use std::time::Duration;
use crate::client::Format;
use super::{Provider, Client, Error, Node, DEFAULT_TIMEOUT_SECS, Consumer};

#[derive(Clone)]
pub struct Context {
    client: Client,
    node: Node,
    id: String,
}

impl Context {
    pub(crate) fn new(client: Client, node: Node, id: String) -> Context {
        Self { client, node, id }
    }

    pub fn consumer(&self, profile: &str) -> Result<Arc<Consumer>, Error> {
        let args = vec![
            self.node.system.id.to_string(),
            self.node.id.clone(),
            self.id.to_string(),
            profile.to_string(),
        ];
        let mut client = self.client.clone();
        let transaction = client.send(Format::Json, "consumers", &args)?;
        let _res = client
            .wait_for_response(transaction, Duration::from_secs(DEFAULT_TIMEOUT_SECS))?;
        Ok(Arc::new(Consumer::new(client, self.clone(), profile.to_string())))
    }

    pub fn provider(&self, profile: &str) -> Result<Arc<Provider>, Error> {
        let args = vec![
            self.node.system.id.to_string(),
            self.node.id.clone(),
            self.id.to_string(),
            profile.to_string(),
        ];
        let mut client = self.client.clone();
        let transaction = client.send(Format::Json, "providers", &args)?;
        let _res = client
            .wait_for_response(transaction, Duration::from_secs(DEFAULT_TIMEOUT_SECS))?;
        Ok(Arc::new(Provider::new(client, self.clone(), profile.to_string())))
    }
}

File: python/src/arete_sdk/context.py
from arete_sdk.consumer import Consumer
from arete_sdk.provider import Provider


class Context:
    def __init__(self, client, node, id):
        self.client = client
        self.node = node
        self.id = id

    def consumer(self, profile):
        args = [self.node.system.id, self.node.id, self.id, profile]
        transaction = self.client.send('json', 'consumers', args)
        self.client.wait_for_response(transaction)
        return Consumer(self.client, self, profile)

    def provider(self, profile):
        args = [self.node.system.id, self.node.id, self.id, profile]
        transaction = self.client.send('json', 'providers', args)
        self.client.wait_for_response(transaction)
        return Provider(self.client, self, profile)
